magma/when.py

```diff
@@ -680,3 +680,71 @@ def otherwise():
     if prev_block is None:
         raise OtherwiseWithoutPrecedingWhenError()
     return prev_block.new_otherwise_block()
+
+
+def _get_builder_ports(builder, names):
+    """Filter out removed ports."""
+    return [getattr(builder, name) for name in names if hasattr(builder, name)]
+
+
+def _check_to_split(value, outputs, to_split):
+    source = value.trace()
+    for elem in source.root_iter():
+        if any(elem is output for output in outputs):
+            to_split.append(source)
+            return
+    if value.has_children() and value.has_elaborated_children():
+        for _, child in value.enumerate_children():
+            _check_to_split(child, outputs, to_split)
+
+
+def _find_values_to_split(builder):
+    """Detect output values that feed into inputs."""
+    to_split = []
+    outputs = _get_builder_ports(builder, builder.output_to_name.values())
+    inputs = _get_builder_ports(builder, builder.input_to_name.values())
+    for value in inputs:
+        _check_to_split(value, outputs, to_split)
+    return to_split
+
+
+def _emit_new_when_assign(value, driver_map, curr_block):
+    """Reconstruct when logic in new set of blocks."""
+    if isinstance(curr_block, _WhenBlock):
+        new_block = when(curr_block._info.condition)
+    elif isinstance(curr_block, _ElseWhenBlock):
+        new_block = elsewhen(curr_block._info.condition)
+    elif isinstance(curr_block, _OtherwiseBlock):
+        new_block = otherwise()
+    with new_block:
+        if curr_block in driver_map:
+            for driver in driver_map[curr_block]:
+                value @= driver
+        for child in curr_block.children():
+            _emit_new_when_assign(value, driver_map, child)
+    for _elsewhen in curr_block.elsewhen_blocks():
+        _emit_new_when_assign(value, driver_map, _elsewhen)
+    if curr_block.otherwise_block:
+        _emit_new_when_assign(value, driver_map, curr_block.otherwise_block)
+
+
+def _build_driver_map(drivee):
+    """
+    driver_map: for each context that drivee is driven, store the driver
+    """
+    driver_map = {}
+    for ctx in drivee._wired_when_contexts:
+        wires = ctx.get_conditional_wires_for_drivee(drivee)
+        driver_map[ctx] = (wire.driver for wire in wires)
+    return driver_map
+
+
+def split_when_cycles(builder, defn):
```

Comment:
if these are all free functions, let's move them to the pass area since theyre not fundamental to the basic when construct